### ISSUE TITLE
Provide a modal-free workaround for assignment submissions

### ISSUE THREAD

**Issue by mtthwhggns** [apluslms/a-plus#669]
We know that modals are breaking the keyboard and screen reader accessibility and as of yet, I've not got close to finding a good solution (we have too many, called in too many places and inconsistently to find a single straightforward solution).

There has been talk of dropping modals altogether, but this is a long way off. 

I just had the idea that we could temporarily work around this in exercises embedded inside chapters by:

- Having a (maybe hidden like the skip links) link right before the exercise, which says something like "open exercise 5 in a new tab".
- This opens the full-page view showing the exercise only (which I believe we have already, but don't link to), where all of the links point to pages, rather than opening modals.
- This page also includes links to submissions and results.

Can anyone think of a problem with this solution?

**Comment by raphendyr** [comment]
I think this should be a viable work-around for now. However, once the exercise UI is fine, the per exercise page should be removed (well, replaced with a JavaScript solution) as mentioned in #665

**Comment by Mankro** [comment]
An interesting idea. There are concerns with it, though.

First some background. In the early A+, we only had exercises and no content chapters ("e-book"). The exercise pages have info boxes in the right side column and the exercise navbar above the exercise is different than in the chapter pages. These right side boxes are never shown in content chapter pages. It is still possible to make courses with only exercises and no chapters. It is also possible to open exercises in the full-page view even if the course embeds the exercises in chapters, but this has been discouraged. 

We have several courses in production that still use the old "only-exercises" format without RST content. Jaakko has been pushing courses using only the RST format with embedded exercises, but I have not been as convinced about that goal.

Most of the time, there is no functional difference between the embedded and full-page view. They look different, which may confuse users. In addition, there are exercises that depend on frontend JavaScript and there is a real concern. The JavaScript behaves differently in chapter pages and full-page exercise pages. 

* chapter pages load exercise descriptions with AJAX after the normal page load. There are some custom JS events used to signal different phases of the loading. The exercise description may include `<script>` elements and those would then activate after the normal page load events have already been triggered.
* full-page exercise pages include all content in the initial page load. JS may wait for the normal page loaded event.
* If I remember correctly, the full-page exercise page does not have the JS event handler for updating the points badge after submissions without reloading the whole page
* I made a hack that exercises could use for initializing their JS without knowing if the exercise is used in chapters or full-page pages. This script shows that structure in the bottom: https://github.com/acos-server/acos-aplus/blob/master/static/resizeiframe.js
* the courses that depend on JS exercises might not work in both exercise pages (chapter and full page) because the teacher most likely tests the course in only one way and does not even recognize this JS problem.

If we link from chapters to the full-page exercises, we have to accept that some exercises do not work correctly. How many users would be impacted by this? Only those who use the keyboard for navigation? Should we add a warning to the exercise page that it might not work in that view (only if the chapter mode is the primary mode of the course).

**Comment by mtthwhggns** [comment]
These are interesting and valid concerns.

> If we link from chapters to the full-page exercises, we have to accept that some exercises do not work correctly. How many users would be impacted by this? Only those who use the keyboard for navigation? Should we add a warning to the exercise page that it might not work in that view (only if the chapter mode is the primary mode of the course).

My concern is that if we leave the dependence on broken modals, users who can't use a mouse and users of screen readers (at least) will be unable to access any exercises. 

If we implement it this way, and make the link visible only on focus, it will be (just like the skip links) quite invisible to other users, so close to zero impact on others (just like the skip links).

> the full-page exercise page does not have the JS event handler for updating the points badge after submissions without reloading the whole page

Forcing the page to reload isn't a bad thing for assistive technology users. It's hard to notify users who are reading a small part of the page that something has changed somewhere else on the page, a refresh signals "the page changed, you should start reading it again". There are live alternatives (look at [ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) if you're curious), but these are not easy to get right. 

> Should we add a warning to the exercise page that it might not work in that view (only if the chapter mode is the primary mode of the course).

I'd be OK with a clear alert at the top of the page in these cases that says something like "this view is intended for users of assistive technology, go to the chapter page if you aren't using any", or whatever, it's still an improvement on where we are now. 

>  In addition, there are exercises that depend on frontend JavaScript and there is a real concern.

Do you know any examples of these?

I just took a quick look through last year's O1, and all of the assignments I've seen so far are using simple form fields (type your answer, choose an answer) or file uploads, so I would guess that these don't depend on any special Javascript. I know that courses like 'Data structures and algorithms" have some custom JS-based exercises, but these have even more potential issues (some exercises themselves are dependent on the mouse, for example).

I found some exercise-only examples using custom JS exercises in another course - these are a valid concern, but they have enough of their own accessibility problems that they probably shouldn't be a concern on their own. 

No solution to this is will be perfect, I hope that this would be enough of an improvement on where we are now that it doesn't totally exclude anyone.

**Comment by raphendyr** [comment]
> Jaakko has been pushing courses using only the RST format with embedded exercises, but I have not been as convinced about that goal.

No, just that exercises should use JavaScript UI and not multiple Python View interface. The content would still be the same and configured the same way. To be fair, I have always tried to ensure we have the HTML+YAML interface and then RST can be used on top of that.

> I made a hack that exercises could use for initializing their JS without knowing if the exercise is used in chapters or full-page pages. This script shows that structure in the bottom: https://github.com/acos-server/acos-aplus/blob/master/static/resizeiframe.js

That seems to be the correct way to initialize a javascript file, so it's not so much a hack than the problem of the browser interface.

And this is kind of the reason, why I would prefer to use the same code, JS Events and interface for all exercise views, so the custom code or what not wouldn't need to be so complicated. It would also ensure that if one view works, then all of them.

**Comment by Mankro** [comment]
I agree with @mtthwhggns that this simple solution would improve accessibility while we are unable to fix the modal dialog issues in the chapter pages. Since JavaScript-based exercises may have plenty of accessibility issues, that is not a good reason for dismissing this solution. This will be an improvement for most cases.

Do users understand the purpose of the link (to the full-page exercise)? Would they think "so I can open the exercise in a new tab, but why bother? I will just use this chapter page"? Do they understand that the modals are broken for assistive technology? Basically, should the link text emphasize that "you really want to open the exercise in a new tab if you are using only keyboard navigation"?

> About JavaScript exercises:
> Do you know any examples of these?

* Theoretical Computer Science (some exercises require the mouse too)
* JSAV in Data Structures and Algorithms (mouse required, I guess screen readers don't work with visual content in graphs and trees so that stops blind students)
* JSVEE (Programming 1 uses these as program execution visualizations without any grading/points)
* relational algebra exercises in Databases use JS for rendering the feedback and the solution editor uses JS
* Acos point-and-click and drag-and-drop exercises especially in English courses (these might even work with the keyboard, but I never tested that)
* most Acos exercise types, jsParsons etc.
* some exercises in Declarative Programming

Somewhat related, the feedback of Scala (Programming 1 and Studio 1 & 2) and Python exercises (using python-grader-utils) often have collapsed sections that don't work without JavaScript. I wonder if those work with keyboard navigation...

**Comment by mtthwhggns** [comment]
> Do users understand the purpose of the link (to the full-page exercise)? Would they think "so I can open the exercise in a new tab, but why bother? I will just use this chapter page"? Do they understand that the modals are broken for assistive technology? Basically, should the link text emphasize that "you really want to open the exercise in a new tab if you are using only keyboard navigation"?

I believe it should. If we have the link hidden until focused with the keyboard, we could consider text like "Open this exercise in a new tab (recommended for keyboard and assistive technology users)" or even "necessary for keyboard and assistive technology users"

It's also worth noting that often users who are used to having problems with assistive technology will check the accessibility statement before digging in too deep. Students worried about being able to access a course also might contact the teacher before enrollment starts. 

Even with the work-around, this should be listed as a known issue in the statement. Something like "it's not possible to access the feedback of in-chapter exercises using the keyboard, but a link to open the exercise in a new tab is provided as a work-around, where the feedback can be accessed using the keyboard".

A more thorough, but (in my opinion) riskier option is completely hiding the exercise from keyboard navigation and screen reader access. This would force the user to follow the link to a different page. It would be annoying for part-time keyboard users who can use a mouse though. My instinctive reaction is that I would not recommend this, but we can discuss it if you like.

> Somewhat related, the feedback of Scala (Programming 1 and Studio 1 & 2) and Python exercises (using python-grader-utils) often have collapsed sections that don't work without JavaScript. I wonder if those work with keyboard navigation...

These work fine, I checked them earlier. To be clear, I was talking about Javascript-heavy exercises (like the Data Structures and Algorithms exercises, which I tested and are mouse-only), not Javascript in general.

**Comment by jaguarfi** [comment]
The PR #735 allows the users who use the keyboard navigation to open the exercises in a new tab (full-page), but I think that we need to redesign the [user workflow](https://www.nngroup.com/articles/wireflows/) for the exercises, and either support [keyboard navigation in modal windows](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html) or remove the modal windows from the workflow.
